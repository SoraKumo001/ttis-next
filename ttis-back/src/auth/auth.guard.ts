import { Injectable, ExecutionContext, createParamDecorator } from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { GqlExecutionContext } from '@nestjs/graphql';

@Injectable()
export class JwtAuthGuard extends AuthGuard('jwt') {
  constructor(){
    super()
  }
  handleRequest(_err, user) {
    return user || null;
  }
  getRequest(context: ExecutionContext) {
    const ctx = GqlExecutionContext.create(context);
    return ctx.getContext().req;
  }
}
export const CurrentUser = createParamDecorator(
  async (_, ctx: ExecutionContext) => ctx.getArgByIndex(2).req.user,
);