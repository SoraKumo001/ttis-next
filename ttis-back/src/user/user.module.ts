import { Module } from '@nestjs/common';
import { UserResolver } from './user.resolver';
import { TypeOrmModule } from '@nestjs/typeorm';
import { User } from './user';
import { UserService } from './user.service';

@Module({
  imports: [TypeOrmModule.forFeature([User])],
  providers: [UserService,UserResolver],
  exports:[UserService]
})
export class UserModule {}
