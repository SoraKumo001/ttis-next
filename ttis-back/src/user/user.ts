import { Entity, PrimaryGeneratedColumn, Column, BaseEntity } from 'typeorm';
import { ObjectType, Field, Int } from '@nestjs/graphql';

@ObjectType()
@Entity()
export class User extends BaseEntity {
  @Field(_ => Int)
  @PrimaryGeneratedColumn() //自動番号
  id!: number;
  @Field()
  @Column({ default: true })
  enable!: boolean;
  @Field()
  @Column({ unique: true })
  name!: string;
  @Column({ nullable: true })
  password?: string;
  @Field()
  @Column({ default: '{}' })
  info!: string;
}
