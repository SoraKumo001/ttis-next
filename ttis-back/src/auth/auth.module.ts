import { Module } from '@nestjs/common';
import { AuthResolver } from './auth.resolver';
import { AuthService } from './auth.service';
import { JwtModule } from '@nestjs/jwt';
import { PassportModule } from '@nestjs/passport';
import { secret } from './secret.secret';
import { UserModule } from '../user/user.module';

@Module({
  imports: [
    UserModule,
    PassportModule,
    JwtModule.register({
      secret: secret,
      signOptions: { expiresIn: '6000s' },
    }),
  ],
  providers: [AuthService, AuthResolver],
})
export class AuthModule {}
