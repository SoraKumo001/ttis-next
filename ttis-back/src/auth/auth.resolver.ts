import {
  Resolver,
  Mutation,
  ObjectType,
  Field,
  Args,
  Query,
} from '@nestjs/graphql';
import { User } from '../user/user';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { getSHA256 } from '@libs/hashTools';

import {
  UseGuards,
} from '@nestjs/common';
import { JwtAuthGuard, CurrentUser } from './auth.guard';
import { AuthService } from './auth.service';

@ObjectType()
export class Login {
  @Field()
  token: string;
  @Field()
  user: User;
}

@Resolver('Auth')
export class AuthResolver {
  constructor(
    private readonly service: AuthService,
  ) {}
  @UseGuards(JwtAuthGuard)
  @Query((_) => User, { nullable: true })
  async currentUser(@CurrentUser() user: User) {
    return user;
  }

  @Mutation((_) => Login, { nullable: true })
  async login(@Args('name') name: string, @Args('password') password: string) {
    return this.service.login(name,password);
  }
}
