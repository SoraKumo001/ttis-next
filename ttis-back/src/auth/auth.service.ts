import { Injectable } from '@nestjs/common';
import { User } from '../user/user';
import { PassportStrategy } from '@nestjs/passport';
import { Strategy, ExtractJwt } from 'passport-jwt';
import { secret } from './secret.secret';
import { JwtService } from '@nestjs/jwt';
import { UserService } from '../user/user.service';

@Injectable()
export class AuthService extends PassportStrategy(Strategy) {
  constructor(
    private readonly userService: UserService,
    private readonly jwt: JwtService,
  ) {
    super({
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
      ignoreExpiration: false,
      secretOrKey: secret,
    });
  }
  async validate({
    id,
    password,
  }: {
    id: number;
    password: string;
  }): Promise<User | undefined> {
    const { userService } = this;
    return userService.validateUser(id, password);
  }
  async login(name: string, password: string) {
    const { userService, jwt } = this;
    const user = await userService.login(name, password);
    if (!user) return null;

    const token = jwt.sign({ id: user.id, password: user.password });
    return { token, user: user };
  }
  async currentUser(user:User) {
    const { jwt } = this;
    if (!user) return null;

    const token = jwt.sign({ id: user.id, password: user.password });
    return { token, user: user };
  }
}
