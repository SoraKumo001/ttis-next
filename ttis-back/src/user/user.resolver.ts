import { Resolver, Args, Mutation, Query, Info, Int } from '@nestjs/graphql';
import { User } from './user';
import { getFields } from '@libs/graphQLTools';
import { UseGuards } from '@nestjs/common';
import { JwtAuthGuard, CurrentUser } from '../auth/auth.guard';
import { UserService } from './user.service';

@Resolver('User')
export class UserResolver {
  constructor(
    private readonly service: UserService,
  ) {}
  @UseGuards(JwtAuthGuard)
  @Query((_) => [User], { nullable: true })
  async users(@Info() info, @CurrentUser() user) {
    if (!user) return null;
    return this.service.users(getFields(info));
  }
  @Mutation((_) => Boolean)
  async deleteUser(
    @Args('id', { type: () => Int }) id: number,
    @CurrentUser() user,
  ) {
    if (!user) return null;
    return this.service.deleteUser(id);
  }

  @Mutation((_) => Boolean)
  async deleteUsers(
    @Args('ids', { type: () => [Int] }) ids: number[],
    @CurrentUser() user,
  ) {
    if (!user) return null;
    return this.service.deleteUsers(ids);
  }

  @Mutation((_) => User, { nullable: true })
  async user(
    @Args('id', { nullable: true, type: () => Int }) id?: number,
    @Args('name', { nullable: true }) name?: string,
    @Args('password', { nullable: true }) password?: string,
  ) {
    return this.service.setUser(id,name,password);
  }
}
