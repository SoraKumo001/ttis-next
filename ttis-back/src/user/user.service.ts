import { Injectable, OnModuleInit } from '@nestjs/common';
import { User } from './user';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { GraphQLFeald } from '@libs/graphQLTools';
import { getSHA256 } from '@libs/hashTools';

@Injectable()
export class UserService implements OnModuleInit {
  constructor(
    @InjectRepository(User)
    private readonly rep: Repository<User>,
  ) {}
  async onModuleInit() {
    const { rep } = this;
    return rep.manager.connection.transaction(async () => {
      return rep.count().then((count) => {
        if (!count) {
          return rep.save({ name: 'admin', password: getSHA256('') });
        }
      });
    });
  }
  async setUser(id?: number, name?: string, password?: string, info?: unknown) {
    const { rep } = this;
    if (!id) {
      if (name && password) {
        return rep.save({
          name,
          password: getSHA256(password),
          info: info && JSON.stringify(info),
        });
      }
    } else {
      const user = await rep.findOne(id);
      if (user) {
        if (name) user.name = name;
        if (password) user.password = getSHA256(password);
        return user.save();
      }
    }
    return null;
  }
  async users(fields?: GraphQLFeald) {
    const { rep } = this;
    return rep.find({
      select: fields as (keyof User)[],
      where: { enable: true },
      order: { id: 'ASC' },
    });
  }
  async deleteUser(id: number) {
    const { rep } = this;
    return !!(await rep.update(id, { enable: false }));
  }

  async deleteUsers(ids: number[]) {
    const { rep } = this;
    return !!(await rep.update(ids, { enable: false }));
  }
  async login(name: string, password: string) {
    const { rep } = this;
    const passwordHash = getSHA256(password);
    return await rep.findOne({
      where: { name, password: passwordHash, enable: true },
    });
  }
  async validateUser(id: number, password: string) {
    const { rep } = this;
    return await rep.findOne({
      where: { id, password, enable: true },
    });
  }
}
