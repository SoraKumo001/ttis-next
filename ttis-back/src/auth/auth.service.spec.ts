import { Test, TestingModule } from '@nestjs/testing';
import { DBModule } from '../db.module';
import { AuthModule } from './auth.module';
import { AuthService } from './auth.service';
import { User } from 'src/user/user';

describe('AuthService', () => {
  let service: AuthService;
  let testUser: User;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [DBModule, AuthModule],
    }).compile();

    service = module.get<AuthService>(AuthService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
  it('Login', async () => {
    expect(
      await service.login('admin', '').then(({ token, user }) => {
        testUser = user;
        return { token: token.substring(0,20), user };
      }),
    ).toMatchSnapshot();
  });
  it('Login Failure', async () => {
    expect(await service.login('admin', 'test')).toMatchSnapshot();
  });
  it('validate', async () => {
    expect(
      await service.validate({ id: testUser.id, password: testUser.password }),
    ).toMatchSnapshot();
  });
});
