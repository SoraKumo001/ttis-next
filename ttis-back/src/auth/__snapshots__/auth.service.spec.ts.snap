// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService Login 1`] = `
Object {
  "token": "eyJhbGciOiJIUzI1NiIs",
  "user": User {
    "enable": true,
    "id": 1,
    "info": "{}",
    "name": "admin",
    "password": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
  },
}
`;

exports[`AuthService Login Failure 1`] = `null`;

exports[`AuthService validate 1`] = `
User {
  "enable": true,
  "id": 1,
  "info": "{}",
  "name": "admin",
  "password": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
}
`;
